### PR TITLE
ENH: Add function to determine time-resolved chpi status

### PR DESCRIPTION
**Pull request by eort**
References issue [#940](https://github.com/mne-tools/mne-bids/issues/940) in the mne-bids repository that is about mne-bids incorrectly setting the cHPI field in the meg recording sidecar. 

Based on that discussion, we (i.e. @larsoner) thought that it might be useful to have a function in mne that returns the time-resolved number of active HPI coils. With that information it should be possible to infer whether continous HPI was active or not. 

This PR is work in progress. I mostly wanted to shift the discussion over here. What still misses:

- [ ] ~~implementation for CTF files~~
- [ ] ~~implementation for KIT files(?)~~
- [x] tests

I don't know anything about CTF files nor any files available (other than those in mne-data), but if I understand the test below correctly, for CTF files it holds that if there are any HPI coils in the data file, they are automatically on. If there are not present, they are obviously off. Can someone who has experience with such files (dis)confirm that? 

https://github.com/mne-tools/mne-python/blob/963453660fc3805a67b924646f4b88f57badcdaa/mne/tests/test_chpi.py#L664

Same holds for KIT files. If it is not clear, we could also implement only for neuromag systems and return some defaults for other systems, together with a info/warning or something like that.

Furthermore, I am not sure whether inference by suffix is robust enough to determine the manufacturer (neuromag, ctf, kit, etc). Is there a better way?

### CODE REVIEW

**Comment by larsoner** [comment]
> If it is not clear, we could also implement only for neuromag systems and return some defaults for other systems, together with a info/warning or something like that.

For now I would make it just work for Neuromag systems and raise an error for others

> I am not sure whether inference by suffix is robust enough to determine the manufacturer (neuromag, ctf, kit, etc). Is there a better way?

The logic in `get_meg_helmet_surf` should be good enough to tell neuromag vs non-neuromag, I'd reuse that (in whatever way it does it) in a DRY way

**Comment by eort** [comment]
Both suggestions sound good. 

I meant to add a test for inactive hpis in neuromag systems, too, but I can't find files in mne_data that I can use for that. The ones I tried miss the `hpi_subsystem` field in their info, so that this line throws a ValueError (`"No appropriate cHPI information found in info["hpi_meas"] and info["hpi_subsystem"]`)
https://github.com/mne-tools/mne-python/blob/01cad560fa1c95ec83dea8cb091b01a331db1b7d/mne/chpi.py#L1410

In my own test file, those fields are present though, such that the method works. Any ideas?

**Comment by larsoner** [comment]
Did you check test_move_anon_raw.fif ? It should have it
```
>>> raw = mne.io.read_raw_fif(mne.datasets.testing.data_path() / 'SSS' / 'test_move_anon_raw.fif', allow_maxshield='yes')
>>> raw.info['hpi_subsystem']
{'ncoil': 5, 'hpi_coils': [{'event_bits': array([256,   0, 256, 256], dtype=int32)}, {'event_bits': array([512,   0, 512, 512], dtype=int32)}, {'event_bits': array([1024,    0, 1024, 1024], dtype=int32)}, {'event_bits': array([2048,    0, 2048, 2048], dtype=int32)}, {'event_bits': array([4096,    0, 4096, 4096], dtype=int32)}], 'event_channel': 'STI201'}
```
for tests that already use it, see https://github.com/mne-tools/mne-python/blob/main/mne/tests/test_chpi.py#L41

**Comment by eort** [comment]
Yeah, I use it to test whether the method works for when the cHPIs were active. But I though I would also test a file where cHPI were inactive. And a file for that I can't find. Or is that overkill?

**Comment by eort** [comment]
yeah, sounds reasonable. As far as I am concerned, the PR is ready for review.

File: mne/chpi.py [hunk 1]
```diff
@@ -1383,3 +1384,40 @@ def _compute_good_distances(hpi_coil_dists, new_pos, dist_limit=0.005):
...
+    Returns
+    -------
+    times : array, shape (n_times)
+        The number of active cHPIs for every timepoint in raw.
+    """
+    # get meg system
+    system, _ = _get_meg_system(raw.info)
+
+    # processing neuromag files
+    if system in ['122m', '306m']:
```
Comment (larsoner) [inline mne/chpi.py:1408]: A better pattern is:
```
if system not in ...:
    raise
chpi_info = ...
```
then the rest of the code follows. That way you save an indentation level

File: mne/chpi.py [hunk 2]
```diff
@@ -1383,3 +1384,40 @@ def _compute_good_distances(hpi_coil_dists, new_pos, dist_limit=0.005):
...
+    Parameters
+    ----------
+    raw : instance of Raw
+        Raw data
+
+            .. versionadded:: 1.2
+
+    Returns
+    -------
+    times : array, shape (n_times)
```
Comment (larsoner) [inline mne/chpi.py:1401]:
```suggestion
    n_active : array, shape (n_times,)
```

File: mne/chpi.py [hunk 3]
```diff
@@ -1383,3 +1384,40 @@ def _compute_good_distances(hpi_coil_dists, new_pos, dist_limit=0.005):
...
+
+    # processing neuromag files
+    if system in ['122m', '306m']:
+        # extract hpi info
+        chpi_info = get_chpi_info(raw.info)
+
+        # extract hpi time series and infer which one was on
+        chpi_ts = raw[chpi_info[1]][0].astype(int)
+        chpi_active = (chpi_ts & chpi_info[2][:, np.newaxis]).astype(bool)
+        times = chpi_active.sum(axis=0)
```
Comment (larsoner) [inline mne/chpi.py:1415]: n_active

**Review by larsoner — COMMENTED** [review]
LGTM other than a couple of minor things! Can you also please

- update `latest.inc` as well
- add it to tutorials/preprocessing/60_maxwell_filtering_sss.py or examples/preprocessing/movement_compensation.py in some suitable place

**Comment by eort** [comment]
I think this is ready for merging @larsoner

File: mne/chpi.py [hunk 4]
```diff
@@ -1383,3 +1384,37 @@ def _compute_good_distances(hpi_coil_dists, new_pos, dist_limit=0.005):
...
+    ----------
+    raw : instance of Raw
+        Raw data
+
+            .. versionadded:: 1.2
+
+    Returns
+    -------
+    n_active : array, shape (n_times)
+        The number of active cHPIs for every timepoint in raw.
```
Comment (larsoner) [inline mne/chpi.py:1402]:
```suggestion

    Returns
    -------
    n_active : array, shape (n_times)
        The number of active cHPIs for every timepoint in raw.

    Notes
    -----
    .. versionadded:: 1.2
```

File: tutorials/preprocessing/60_maxwell_filtering_sss.py [hunk 5]
```diff
@@ -296,7 +296,20 @@
...
+# want to doublecheck, you can use :func:`mne.chpi.get_active_chpi` (currently
+# only implemented for neuromag systems).
+
+chpi_fif_file = os.path.join(mne.datasets.testing.data_path(), 'SSS',
+                             'test_move_anon_raw.fif')
+raw = mne.io.read_raw_fif(chpi_fif_file, allow_maxshield='yes')
+
+# time-resolved information on active HPI coils
+# if all hpi were inactive n_active is a zero-array
+n_active = mne.chpi.get_active_chpi(raw)
```
Comment (larsoner) [inline tutorials/preprocessing/60_maxwell_filtering_sss.py:309]:
```suggestion
n_active = mne.chpi.get_active_chpi(raw)
print(f'Average number of coils active during recording: {n_active.mean()}')
```

**Comment by larsoner** [comment]
Thanks @eort !

**Pull request merged**